feat: add release utilities

bdura opened the pull request:
## Description

This PR ports release utilities from spaCy. These include:
- shell scripts within the `bin/` directory
- automatic git commit extraction within `setup.py`

I also added a step that extracts the library version from `setup.cfg`, to avoid repetition.

svlandeg reviewed (approved):
Nice! It wasn't easy to find a nit-pick, but I managed :trollface: 

Feel free to merge though!

bdura commented on licenses/3rd_party_licenses.txt:
```diff
@@ -0,0 +1,38 @@
+Third Party Licenses for Weasel
+==============================
```
😚 👌

bdura commented on setup.cfg:
```diff
@@ -1,4 +1,6 @@
 [metadata]
+name = weasel
+version = 0.1.0
```
Well, it is an accepted practice: https://semver.org/#how-should-i-deal-with-revisions-in-the-0yz-initial-development-phase

But it's inconsistent with our other projects, I'll change that.

adrianeboyd commented:
The git version is related to changes in `spacy` while you're running some project, not changes in the project CLI itself, if that makes sense? (And then no third party code/licenses, which seems like it would be best to avoid for this.)

bdura commented:
Oh, I guess we can use `importlib.metadata` here as well.

adrianeboyd commented:
Can you explain to me why you need `__version__` for a lock file?

bdura commented:
> Can you explain to me why you need `__version__` for a lock file?

The tag is a misnomer, but: https://github.com/explosion/weasel/blob/main/weasel/cli/run.py#L306

By the way, the version is _always_ checked. See [here](https://github.com/explosion/weasel/blob/main/weasel/cli/run.py#L227) and [there](https://github.com/explosion/weasel/blob/main/weasel/cli/run.py#L108): the argument is not used and always defaults to `True`.

adrianeboyd commented:
Ah, then also use `importlib.metadata` in the same way to get the version for this?

But I'm not actually sure what we want to do about the spacy version in the lock files in general...

bdura commented:
Yes, we should at the very least add a `weasel_version` key, and probably rename these to something else entirely for that matter, in order to avoid confusion between that and the corresponding config keys.

There are still many instances of `spacy` lurking within the code, I'm gradually making my way through.

bdura commented:
I think we should change the default behaviour to _not_ re-run on a version change, by the way. Now that the project is separate from spaCy itself, I'm not sure it makes sense to re-run a computationally costly experiment should minor/patch version of the experiment manager change.

bdura commented:
Converting this to draft until #15 is merged, to avoid to many conflicts.

bdura commented:
Ok, that PR turned out huge - sorry about that.

I removed the git commit and version checks altogether, and added a check that the deprecated environment variables are not set. Let me know.

adrianeboyd commented on weasel/about.py:
```diff
@@ -1,3 +1,2 @@
-__version__ = "3.5.0"
 __projects__ = "https://github.com/explosion/projects"
 __projects_branch__ = "v3"
```
This will be in spacy instead, let's just remove this whole file?

bdura replied:
It's used for the `clone` command... We should probably remove it nonetheless, but let's do that in a separate PR?

adrianeboyd commented on weasel/compat.py:
```diff
@@ -3,3 +3,8 @@
 is_windows = sys.platform.startswith("win")
 is_linux = sys.platform.startswith("linux")
 is_osx = sys.platform == "darwin"
+
+try:
+    from importlib.metadata import version  # noqa
+except ImportError:
+    from catalogue._importlib_metadata import version  # type: ignore [no-redef]  # noqa
```
If we're going to delay this to v4, then we can drop python 3.6 and 3.7 and we don't need this compat section.

bdura replied:
Done, thanks!
